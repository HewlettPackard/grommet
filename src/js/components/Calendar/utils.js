
// Utility functions for the Calendar.
// Just what's needed to avoid having to include a dependency like momentjs.

const DAY_MILLISECONDS = 24 * 60 * 60 * 1000;

export const addDays = (date, days) =>
  (new Date(date.getTime() + (DAY_MILLISECONDS * days)));

export const subtractDays = (date, days) => addDays(date, -days);

export const addMonths = (date, months) => {
  const result = new Date(date);
  const years = Math.floor((date.getMonth() + months) / 12);
  result.setFullYear(date.getFullYear() + years);
  result.setMonth((date.getMonth() + months) % 12);
  return result;
};

export const subtractMonths = (date, months) => addMonths(date, -months);

export const sameDay = (date1, date2) => (
  date1.getFullYear() === date2.getFullYear() &&
  date1.getMonth() === date2.getMonth() &&
  date1.getDate() === date2.getDate()
);

export const sameDayOrAfter = (date1, date2) =>
  (date1.getFullYear() > date2.getFullYear() ||
    (date1.getFullYear() === date2.getFullYear() && (
      date1.getMonth() > date2.getMonth() || (
        date1.getMonth() === date2.getMonth() && (date1.getDate() >= date2.getDate())
      )
    ))
  );

export const sameDayOrBefore = (date1, date2) =>
  (date1.getFullYear() < date2.getFullYear() ||
    (date1.getFullYear() === date2.getFullYear() && (
      date1.getMonth() < date2.getMonth() || (
        date1.getMonth() === date2.getMonth() && (date1.getDate() <= date2.getDate())
      )
    ))
  );

export const daysApart = (date1, date2) =>
  Math.floor((date1.getTime() - date2.getTime()) / DAY_MILLISECONDS);

// betweenDates takes and array of two elements and checks if the
// supplied date lies between them, inclusive.
// returns 2 if exact match to one end, 1 if between, undefined otherwise
export const betweenDates = (date, dates) => {
  let result;
  if (dates) {
    const [from, to] = dates.map(d => new Date(d));
    if (sameDay(date, from) || sameDay(date, to)) {
      result = 2;
    } else if (sameDayOrAfter(date, from) && sameDayOrBefore(date, to)) {
      result = 1;
    }
  } else {
    result = 1;
  }
  return result;
};

// withinDates takes and array of string dates or 2 element arrays and
// checks whether the supplied date matches any string or is between
// any dates in arrays.
// returns 2 if exact match, 1 if between, undefined otherwise
export const withinDates = (date, dates) => {
  let result;
  if (dates) {
    if (Array.isArray(dates)) {
      dates.some((d) => {
        if (typeof d === 'string') {
          if (sameDay(date, new Date(d))) {
            result = 2;
          }
        } else {
          result = betweenDates(date, d);
        }
        return result;
      });
    } else if (sameDay(date, new Date(dates))) {
      result = 2;
    }
  }
  return result;
};
