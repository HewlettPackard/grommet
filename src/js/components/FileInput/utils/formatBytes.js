const SI_CONVERSION_FACTOR = 1000;
const IEC_CONVERSION_FACTOR = 1024;

const getBrowser = () => {
  const currentOS = ['Win', 'Linux', 'Mac'].find(
    (v) => window.navigator.userAgent.indexOf(v) >= 0,
  );

  return currentOS;
};

const defaultFormat = (size) => {
  const units = ['B', 'KB', 'MB', 'GB', 'TB'];
  const factor = SI_CONVERSION_FACTOR;
  let index = 0;
  let num = size;
  while (num >= factor && index < units.length - 1) {
    num /= factor;
    index += 1;
  }
  return `${num.toFixed(1)} ${units[index]}`;
};

const windowsFormat = (size) => {
  const num = Math.ceil(size / IEC_CONVERSION_FACTOR);

  function formatNumber(number) {
    return number.toString().replace(/(\d)(?=(\d{3})+(?!\d))/g, '$1.');
  }

  return `${formatNumber(num)} ${'KB'}`;
};

const makeFormatBytes = (browser) => (size) => {
  switch (browser) {
    case 'Win':
      return windowsFormat(size);
    default:
      return defaultFormat(size);
  }
};

const formatBytes = makeFormatBytes(getBrowser());

export { formatBytes };
