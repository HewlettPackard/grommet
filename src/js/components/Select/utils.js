import React, { forwardRef } from 'react';
import styled from 'styled-components';
import { Box } from '../Box';
import { Button } from '../Button';
import { DropButton } from '../DropButton';
import { Text } from '../Text';
import { TextInput } from '../TextInput';
import {
  getHoverIndicatorStyle,
  selectedStyle,
  controlBorderStyle,
  normalizeColor,
} from '../../utils';

export const applyKey = (option, key) => {
  if (option === undefined) return undefined;
  if (typeof key === 'object') return applyKey(option, key.key);
  if (typeof key === 'function') return key(option);
  if (key !== undefined) return option[key];
  return option;
};

// position relative is so scroll can be managed correctly
export const OptionsBox = styled.div`
  position: relative;
  scroll-behavior: smooth;
  overflow: auto;
  outline: none;
`;

export const HiddenInput = styled.input`
  display: none;
`;

export const SelectOption = styled(Button)`
  ${(props) => props.selected && props.textComponent && selectedStyle}
  // applies theme.global.hover.background to the active
  // option for mouse and keyboard interactions
  ${(props) =>
    props.active &&
    getHoverIndicatorStyle(
      !props.children && !props.theme.select.options ? undefined : 'background',
      props.theme,
    )}
  display: block;
  width: 100%;
`;

export const SelectTextInput = styled(TextInput)`
  cursor: ${(props) => (props.defaultCursor ? 'default' : 'pointer')};
`;

export const StyledSelectDropButton = styled(DropButton)`
  ${(props) => !props.callerPlain && controlBorderStyle};
  ${(props) =>
    props.theme.select &&
    props.theme.select?.control &&
    props.theme.select?.control?.extend};
  ${(props) => props.open && props.theme.select?.control?.open};
`;

export const DefaultSelectTextInput = forwardRef(
  (
    { a11yTitle, disabled, id, name, placeholder, value, size, theme, ...rest },
    ref,
  ) => (
    <SelectTextInput
      a11yTitle={a11yTitle}
      // When Select is disabled, we want to show a default cursor
      // but not have disabled styling come from TextInput
      // Disabled can be a bool or an array of options to disable.
      // We only want to disable the TextInput if the control
      // button should be disabled which occurs when disabled
      // equals true.
      defaultCursor={disabled === true || undefined}
      focusIndicator={false}
      id={id ? `${id}__input` : undefined}
      name={name}
      ref={ref}
      {...rest}
      tabIndex="-1"
      type="text"
      placeholder={placeholder}
      plain
      readOnly
      value={value}
      size={size}
      theme={theme}
    />
  ),
);

export const EmptySearchOption = ({
  emptySearchMessage,
  selectOptionsStyle,
  theme,
}) => (
  <SelectOption
    key="search_empty"
    tabIndex="0"
    role="menuitem"
    hoverIndicator="background"
    disabled
    aria-live="polite"
  >
    <Box {...selectOptionsStyle}>
      <Text {...theme.select.container.text}>{emptySearchMessage}</Text>
    </Box>
  </SelectOption>
);

export const getOptionLabel = (index, options, labelKey) =>
  applyKey(options[index], labelKey);

export const getOptionValue = (index, options, valueKey) =>
  applyKey(options[index], valueKey);

export const checkDisabled = (
  index,
  disabled,
  disabledKey,
  options,
  valueKey,
) => {
  const option = options[index];
  let result;
  if (disabledKey) {
    result = applyKey(option, disabledKey);
  } else if (Array.isArray(disabled)) {
    if (typeof disabled[0] === 'number') {
      result = disabled.indexOf(index) !== -1;
    } else {
      const optionVal = getOptionValue(index, options, valueKey);
      result = disabled.indexOf(optionVal) !== -1;
    }
  }
  return result;
};

export const calcValuedValue = (value, valueKey) => {
  if (Array.isArray(value))
    return value.map((v) =>
      valueKey && valueKey.reduce ? v : applyKey(v, valueKey),
    );
  return valueKey && valueKey.reduce ? value : applyKey(value, valueKey);
};

export const changeEvent = (inputRef, nextValue) => {
  // Calling set value function directly on input because React library
  // overrides setter `event.target.value =` and loses original event
  // target fidelity.
  // https://stackoverflow.com/a/46012210
  const nativeInputValueSetter = Object.getOwnPropertyDescriptor(
    window.HTMLInputElement.prototype,
    'value',
  ).set;
  nativeInputValueSetter.call(inputRef.current, nextValue);
  const event = new Event('input', { bubbles: true });
  inputRef.current.dispatchEvent(event);
};

export const getSelectIcon = (icon, theme, open) => {
  let SelectIcon;
  switch (icon) {
    case false:
      break;
    case true:
    case undefined:
      SelectIcon =
        open && theme.select.icons.up
          ? theme.select.icons.up
          : theme.select.icons.down;
      break;
    default:
      SelectIcon = icon;
  }
  return SelectIcon;
};

// if labelKey is a function and valueLabel is not defined
// we should use the labelKey function to display the
// selected value
export const getDisplayLabelKey = (
  labelKey,
  allOptions,
  optionIndexesInValue,
  selectValue,
) => {
  const optionLabelKey = applyKey(
    allOptions[optionIndexesInValue[0]],
    labelKey,
  );
  if (
    !selectValue &&
    optionIndexesInValue.length === 1 &&
    typeof optionLabelKey === 'object'
  )
    return optionLabelKey;
  return undefined;
};

export const getIconColor = (theme) =>
  normalizeColor(theme.select.icons.color || 'control', theme);
