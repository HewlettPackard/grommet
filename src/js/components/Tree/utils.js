export const setupNodes = nodes =>
  nodes.map(node => ({
    name: node.name,
    childs: setupNodes(node.childs || []),
    opened: false,
  }));

export const isSameNode = (node1, node2) => node1 === node2;

export const changeNode = (positions, nodes) => {
  const [currentPosition, ...otherPositions] = positions;
  const currentNode = nodes[currentPosition];

  if (otherPositions.length) {
    return nodes.map(node =>
      isSameNode(node, currentNode)
        ? {
            ...node,
            childs: changeNode(otherPositions, node.childs),
          }
        : { ...node },
    );
  }

  return nodes.map(node =>
    isSameNode(node, currentNode)
      ? {
          ...node,
          opened: !node.opened,
        }
      : { ...node },
  );
};
