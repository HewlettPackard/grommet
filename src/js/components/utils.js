import { css } from 'styled-components';

export const focusStyle = css`
  border-color: ${
    props => (
      props.theme.global.focus.border.color ||
      props.theme.global.colors.accent[0]
    )
  };
  box-shadow: 0 0 1px 1px ${
    props => (
      props.theme.global.focus.border.color ||
      props.theme.global.colors.accent[0]
    )
  };
`;

export function parseMetricToInt(fontAsString) {
  return parseInt(fontAsString.replace(/[^0-9]/g, ''), 10);
}

export function findAllByType(component, type) {
  let matches = [];

  if (component.type === type) {
    matches.push(component);
  }

  if (component.children) {
    component.children.forEach((child) => {
      matches = matches.concat(findAllByType(child, type));
    });
  }

  return matches;
}

export default { findAllByType, focusStyle, parseMetricToInt };
