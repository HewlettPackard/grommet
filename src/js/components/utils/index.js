export * from './styles';
export * from './mixins';
export * from './DOM';
export { default as Drop } from './Drop';
