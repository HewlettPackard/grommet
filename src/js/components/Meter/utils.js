import { colorForName } from '../../utils';

export const strokeProps = (color, theme) => {
  const result = {};
  if (color) {
    if (typeof color === 'object') {
      result.stroke = colorForName(color.color, theme);
      if (color.opacity) {
        result.strokeOpacity = (color.opacity === true ?
          theme.global.opacity.medium : theme.global.opacity[color.opacity]);
      }
    } else {
      result.stroke = colorForName(color, theme);
    }
  }
  return result;
};

export default { strokeProps };
