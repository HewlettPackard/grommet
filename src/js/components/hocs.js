import React, { Component } from 'react';
import PropTypes from 'prop-types';
import getDisplayName from 'recompose/getDisplayName';

import baseTheme from '../themes/vanilla';
import { deepMerge } from '../utils';

export const withFocus = (WrappedComponent) => {
  class FocusableComponent extends Component {
    state = {
      mouseActive: false,
      focus: false,
    }
    setMouseActive() {
      this.setState({ mouseActive: true });
    }
    resetMouseActive() {
      this.setState({ mouseActive: false });
    }
    setFocus() {
      const { mouseActive } = this.state;
      if (mouseActive === false) {
        this.setState({ focus: true });
      }
    }
    resetFocus() {
      this.setState({ focus: false });
    }
    render() {
      const { focus } = this.state;
      return (
        <WrappedComponent
          focus={focus}
          {...this.props}
          onMouseDown={(event) => {
            this.setMouseActive();
            const { onMouseDown } = this.props;
            if (onMouseDown) {
              onMouseDown(event);
            }
          }}
          onMouseUp={(event) => {
            this.resetMouseActive();
            const { onMouseUp } = this.props;
            if (onMouseUp) {
              onMouseUp(event);
            }
          }}
          onFocus={(event) => {
            this.setFocus();
            const { onFocus } = this.props;
            if (onFocus) {
              onFocus(event);
            }
          }}
          onBlur={(event) => {
            this.resetFocus();
            const { onBlur } = this.props;
            if (onBlur) {
              onBlur(event);
            }
          }}
        />
      );
    }
  }

  FocusableComponent.displayName = getDisplayName(WrappedComponent);

  return FocusableComponent;
};

export const withTheme = (WrappedComponent) => {
  class ThemedComponent extends Component {
    static contextTypes = {
      theme: PropTypes.object,
    }
    render() {
      const { theme, ...rest } = this.props;
      const { theme: contextTheme } = this.context;
      const localTheme = deepMerge(baseTheme, contextTheme, theme);
      return (
        <WrappedComponent theme={localTheme} {...rest} />
      );
    }
  }

  ThemedComponent.displayName = getDisplayName(WrappedComponent);

  return ThemedComponent;
};

export default { withFocus, withTheme };
