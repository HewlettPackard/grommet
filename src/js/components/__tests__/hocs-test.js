import React from 'react';
import renderer from 'react-test-renderer';

import { findAllByType } from '../../utils';

import { withFocus } from '../hocs';

const Test = withFocus(({ focus, ...rest }) => (
  <div {...rest}>
    { focus ? 'focus' : 'no focus' }
  </div>
));

test('withFocus set focus', () => {
  const component = renderer.create(<Test />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  const container = findAllByType(tree, 'div');
  container[0].props.onFocus();

  tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('withFocus skip focus when mouse active', () => {
  const component = renderer.create(<Test />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  const container = findAllByType(tree, 'div');
  container[0].props.onMouseDown();
  container[0].props.onFocus();

  tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  container[0].props.onMouseUp();
  container[0].props.onFocus();

  tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  container[0].props.onBlur();

  tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('withFocus calls callback', () => {
  const onFocus = jest.fn();
  const onBlur = jest.fn();
  const onMouseDown = jest.fn();
  const onMouseUp = jest.fn();
  const component = renderer.create(
    <Test
      onFocus={onFocus}
      onBlur={onBlur}
      onMouseDown={onMouseDown}
      onMouseUp={onMouseUp}
    />
  );
  const tree = component.toJSON();

  const container = findAllByType(tree, 'div');
  container[0].props.onFocus();
  container[0].props.onBlur();
  container[0].props.onMouseDown();
  container[0].props.onMouseUp();

  expect(onFocus).toBeCalled();
  expect(onBlur).toBeCalled();
  expect(onMouseDown).toBeCalled();
  expect(onMouseUp).toBeCalled();
});

