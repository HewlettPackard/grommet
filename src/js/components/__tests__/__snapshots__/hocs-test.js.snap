// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withFocus set focus 1`] = `
<div
  onBlur={[Function]}
  onFocus={[Function]}
>
  no focus
</div>
`;

exports[`withFocus set focus 2`] = `
<div
  onBlur={[Function]}
  onFocus={[Function]}
>
  no focus
</div>
`;

exports[`withFocus skip focus when mouse active 1`] = `
<div
  onBlur={[Function]}
  onFocus={[Function]}
>
  no focus
</div>
`;

exports[`withFocus skip focus when mouse active 2`] = `
<div
  onBlur={[Function]}
  onFocus={[Function]}
>
  no focus
</div>
`;

exports[`withFocus skip focus when mouse active 3`] = `
<div
  onBlur={[Function]}
  onFocus={[Function]}
>
  no focus
</div>
`;

exports[`withFocus skip focus when mouse active 4`] = `
<div
  onBlur={[Function]}
  onFocus={[Function]}
>
  no focus
</div>
`;
